test/unit/app/controllers/permissions/permissions-controller-test.js

```diff
@@ -0,0 +1,1392 @@
+          )
+
+          const accounts = await permController.getAccounts(ORIGINS.a)
+          assert.deepEqual(
+            accounts, [], 'origin should have no accounts'
+          )
+        })
+
+      const accounts = await permController.getAccounts(ORIGINS.a)
+
+      assert.deepEqual(accounts, [], 'origin should not have any accounts')
+
+      assert.equal(
+        permController.pendingApprovals.size, 1,
+        'perm controller should have single pending approval',
+      )
+
+      const id = permController.pendingApprovals.keys().next().value
+
+      permController.rejectPermissionsRequest(id)
+    })
+
+    it('throws on bad request', async function () {
+
+      assert.rejects(
+        () => permController._requestPermissions(
+          ORIGINS.a, PERMS.requests.does_not_exist()
+        ),
+        'should reject bad request'
+      )
+    })
+  })
+
+  // see permissions-middleware-test for testing the middleware itself
+  describe('createMiddleware', function () {
+
+    let permController
+
+    beforeEach(function () {
+      permController = initPermController()
+    })
+
+    it('should throw on bad origin', function () {
+
+      assert.throws(
+        () => permController.createMiddleware({ origin: {} }),
+        ERRORS.createMiddleware.badOrigin()
+      )
+
+      assert.throws(
+        () => permController.createMiddleware({ origin: '' }),
+        ERRORS.createMiddleware.badOrigin(),
+      )
+
+      assert.throws(
+        () => permController.createMiddleware({}),
+        ERRORS.createMiddleware.badOrigin(),
+      )
+    })
+
+    it('should create a middleware', function () {
+
+      let middleware
+      assert.doesNotThrow(
+        () => {
+          middleware = permController.createMiddleware({ origin: ORIGINS.a })
+        },
+        'should not throw'
+      )
+
+      assert.equal(
+        typeof middleware, 'function',
+        'should return function'
+      )
+
+      assert.equal(
+        middleware.name, 'engineAsMiddleware',
+        'function name should be "engineAsMiddleware'
+      )
+    })
+
+    it('should create a middleware with extensionId', function () {
+
+      const extensionId = 'fooExtension'
+
+      let middleware
+      assert.doesNotThrow(
+        () => {
+          middleware = permController.createMiddleware({
+            origin: ORIGINS.a,
+            extensionId,
+          })
+        },
+        'should not throw'
+      )
+
+      assert.equal(
+        typeof middleware, 'function',
+        'should return function'
+      )
+
+      assert.equal(
+        middleware.name, 'engineAsMiddleware',
+        'function name should be "engineAsMiddleware'
+      )
+
+      const metadataStore = permController.store.getState()[METADATA_STORE_KEY]
+
+      assert.deepEqual(
+        metadataStore[ORIGINS.a], { extensionId },
+        'metadata should be stored'
+      )
+    })
+  })
+
+  describe('adding and removing pending approvals', function () {
```

Comment:
These also seem to be internal functions, so I'm not sure there's value in testing them. If a path isn't exposed via a public method, that's probably a sign that the path isn't needed, or the tests of the public methods need to be expanded.

Reply:
For discussion, see: https://github.com/MetaMask/metamask-extension/pull/7969#discussion_r390700843